org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+            if (SSL.ConfigJSSDefaultCertAuthCallback(ssl_fd) == SSL.SECFailure) {
+                throw new RuntimeException("Unable to configure JSSNativeTrustManager on this JSSengine: " + errorText(PR.GetError()));
+            }
+        }
+    }
+
+    private void createLoggingSocket() {
+        if (debug_port == 0) {
+            return;
+        }
+
+        try {
+            ss_socket = new ServerSocket(debug_port);
+            ss_socket.setReuseAddress(true);
+            ss_socket.setReceiveBufferSize(BUFFER_SIZE);
+
+            c_socket = new Socket(ss_socket.getInetAddress(), ss_socket.getLocalPort());
+            c_socket.setReuseAddress(true);
+            c_socket.setReceiveBufferSize(BUFFER_SIZE);
+            c_socket.setSendBufferSize(BUFFER_SIZE);
+
+            s_socket = ss_socket.accept();
+            s_socket.setReuseAddress(true);
+            s_socket.setReceiveBufferSize(BUFFER_SIZE);
+            s_socket.setSendBufferSize(BUFFER_SIZE);
+
+            s_ostream = s_socket.getOutputStream();
+
+            c_ostream = c_socket.getOutputStream();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to enable debug socket logging! " + e.getMessage(), e);
+        }
+    }
+
+    public void beginHandshake() {
+        debug("JSSEngine: beginHandshake()");
+
+        // We assume beginHandshake(...) is the entry point for initializing
+        // the buffer. In particular, wrap(...) / unwrap(...) *MUST* call
+        // beginHandshake(...) if ssl_fd == null.
+
+        // ssl_fd == null <-> we've not initialized anything yet.
+        if (ssl_fd == null) {
+            init();
+        }
+
+        // Always, reset the handshake status, using the existing
+        // socket and configuration (which might've been just created).
+        if (SSL.ResetHandshake(ssl_fd, as_server) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to begin handshake: " + errorText(PR.GetError()));
+        }
+    }
+
+    public void closeInbound() {
+        debug("JSSEngine: closeInbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_RCV);
+        is_inbound_closed = true;
+    }
+
+    public void closeOutbound() {
+        debug("JSSEngine: closeOutbound()");
+
+        PR.Shutdown(ssl_fd, PR.SHUTDOWN_SEND);
+        is_outbound_closed = true;
+    }
+
+    public String getHostname() {
+        return hostname;
+    }
+
+    public Runnable getDelegatedTask() {
+        debug("JSSEngine: getDelegatedTask()");
+
+        // We fake being a non-blocking SSLEngine. In particular, we never
+        // export tasks as delegated tasks (e.g., OCSP checking), so this
+        // method will always return null.
+
+        return null;
+    }
+
+    public SSLEngineResult.HandshakeStatus getHandshakeStatus() {
+        debug("JSSEngine: getHandshakeStatus()");
+        try {
+            updateHandshakeState();
+        } catch (SSLException se) {
```

Comment:
This is described in the [documentation on handshaking](https://github.com/cipherboy/jss/blob/jss-engine-docs/docs/usage/jssengine.md#handshaking).

(It is necessary to keep stepping the unknown handshake state and ensure that it gets updated frequently enough -- see for instance, [the tests](https://github.com/cipherboy/jss/blob/ssl-engine/org/mozilla/jss/tests/TestSSLEngine.java#L85-L86) where we want to make sure we keep calling `getHandshakeStatus(...)` in order that the state gets updated, even if we're in an unknown state). 

@jmagne -- Is the comment in the `catch` block not sufficient?

Reply:
I suppose I could make `updateHandshakeState()` not throw an exception at all, since we're never passing it back up, would that be better?

Reply:
I've gone ahead and done that. I think that makes the intent a little clearer. However, this try-catch still needs to remain for reasons documented below. Other try/catches are removed.

Reply:
As discussed on IRC, removing the try-catch block is not expected to have a negative impact. My preference is to remove the try-catch block since it could accidentally hide important exceptions. In general we should only add a try-catch block only if we want to explicitly handle the exception.

Reply:
Block is removed.